mlflow/pyfunc/__init__.py

```diff
@@ -1054,6 +1127,46 @@ def spark_udf(spark, model_uri, result_type="double", env_manager=_EnvManager.LO
 
     model_metadata = Model.load(os.path.join(local_model_path, MLMODEL_FILE_NAME))
 
+    model_output_schema = model_metadata.get_output_schema()
+
+    if result_type is None:
+        if model_output_schema is None:
+            _logger.warning(
+                "No 'result_type' provided for spark_udf and the model metadata does not "
+                "have an output schema. 'result_type' is set to 'double' type."
+            )
+            result_type = "double"
+        else:
+            result_type = _infer_spark_udf_return_type(model_output_schema)
+    else:
+        result_type = "boolean" if result_type == "bool" else result_type
+
+    if not isinstance(result_type, SparkDataType):
+        result_type = _parse_datatype_string(result_type)
+
+    elem_type = result_type
+    if isinstance(elem_type, ArrayType):
+        elem_type = elem_type.elementType
+        if isinstance(elem_type, ArrayType):
+            elem_type = elem_type.elementType
```

Comment:
Can we add an inline comment explaining what's going on here?

Reply:
Related code is moved to `_check_udf_return_type(result_type)`, I will remove the logic here.